### PR TITLE
Add Che client with token cache

### PR DESCRIPTION
**Pull request by rajivnathan**
Adds a Che client to the member operator that does the following:

- looks up the Che Keycloak server
- uses credentials from the member-operator-secret to request a token from Che's Keycloak server
- retrieves the token from the server and caches it
- reuses the cached token until it expires

More functions will be added to che.go in PRs to come. (eg. Find user and Delete user functions)

Related issue: https://issues.redhat.com/browse/CRT-846

### CODE REVIEW

File: pkg/che/tokencache.go [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+
+func newTokenCache() *tokenCache {
+	return &tokenCache{
+		httpClient: newHTTPClient(),
+	}
+}
+
+// getToken returns the token needed to use Che user APIs, or an error if there was a problem getting the token
+func (tc *tokenCache) getToken(cl client.Client, cfg *crtCfg.Config) (TokenSet, error) {
+	defer tc.RUnlock()
```
Comment (alexeykazakov) [inline pkg/che/tokencache.go:40]: This function not just reads the token. It also writes it in line 62: `obtainNewToken()`. So you can't just use RLock here.
You can use something like this: https://github.com/codeready-toolchain/devcluster/blob/master/pkg/ibmcloud/cluster.go#L64

File: pkg/che/che.go [hunk 2]
```diff
@@ -0,0 +1,48 @@
+package che
+
+import (
+	"context"
+	"fmt"
+
+	crtConfig "github.com/codeready-toolchain/member-operator/pkg/configuration"
```
[the remark this answers is not in the document]
Comment (rajivnathan) [reply pkg/che/che.go]: 🤦  - I know this, but somehow I keep making this mistake...hopefully this is the last time

fixed in 45affeaf57ecb3b1211ca474e81310e24762050e

File: pkg/che/tokencache.go [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+	"fmt"
+	"io"
+	"net"
+	"net/http"
+	"net/url"
+	"strings"
+	"sync"
+	"time"
+
+	crtCfg "github.com/codeready-toolchain/member-operator/pkg/configuration"
```
[the remark this answers is not in the document]
Comment (rajivnathan) [reply pkg/che/tokencache.go]: fixed in 45affeaf57ecb3b1211ca474e81310e24762050e

File: pkg/che/tokencache.go (hunk 1 above)
Comment (rajivnathan) [reply pkg/che/tokencache.go:40]: oops, I actually meant to use `Unlock` and `Lock` not `RUnlock` and `RLock`. Let me try something like https://github.com/codeready-toolchain/devcluster/blob/master/pkg/ibmcloud/cluster.go#L64. I did briefly try that but ran into some issue, but let me try again.

Comment (rajivnathan) [reply pkg/che/tokencache.go]: Fixed in 24cf6c55f1de637ac561cbd681c6896adb87fa6a

File: pkg/che/che.go [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+// InitDefaultCheClient initializes the default Che service instance
+func InitDefaultCheClient(cfg *crtConfig.Config, cl client.Client) {
+	DefaultClient = &Client{
+		config:     cfg,
+		k8sClient:  cl,
+		tokenCache: newTokenCache(),
+	}
+}
+
+func getRouteURL(cl client.Client, namespace, name string) (string, error) {
```
[the remark this answers is not in the document]
Comment (MatousJobanek) [reply pkg/che/che.go:36]: In addition, it looks like a copy of https://github.com/codeready-toolchain/member-operator/blob/master/pkg/controller/memberstatus/memberstatus_controller.go#L375-L387 - wouldn't it be good to export it somewhere and reuse it so we don't have to maintain it at two places?

File: pkg/che/tokencache.go [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+		return TokenSet{}, err
+	}
+	log.Info("Che Keycloak Route", "URL", cheKeycloakURL)
+
+	// get che user token
+	if err = tc.obtainNewToken(cheKeycloakURL+tokenPath, user, pass); err != nil {
+		return TokenSet{}, err
+	}
+
+	return *tc.token, nil
```
Comment (MatousJobanek) [inline pkg/che/tokencache.go:69]: I would rather move it to separated method so the read and write locks are more separated, but it's just my point of view ;-)

File: pkg/che/tokencache.go [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+	err := json.Unmarshal([]byte(jsonString), &token)
+	if err != nil {
+		return nil, errors.Wrapf(err, "error when unmarshal json with access token %s ", jsonString)
+	}
+	return &token, nil
+}
+
+// tokenExpired return false if the token is not nil and good for at least one more minute
+func tokenExpired(token *TokenSet) bool {
+	return token == nil || time.Now().After(time.Unix(token.Expiration-60, 0))
```
Comment (MatousJobanek) [inline pkg/che/tokencache.go:139]: What is the default expiration time of the token?

File: pkg/rest/rest.go [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+	buf := new(bytes.Buffer)
+	buf.ReadFrom(body)
+	return buf.String()
+}
+
+// CloseResponse reads the body and close the response. To be used to prevent file descriptor leaks.
+func CloseResponse(response *http.Response) {
+	ioutil.ReadAll(response.Body)
+	response.Body.Close()
+}
```
Comment (MatousJobanek) [inline pkg/rest/rest.go:21]: you don't handle the errors at all here

**Review by MatousJobanek — COMMENTED** [review]
Good job :+1: 
Please, don't forget to change the Secret in toolchain-infra

File: pkg/che/che.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (rajivnathan) [reply pkg/che/che.go:36]: I'll be using it for che API calls in che.go as well but I agree it can be a common function somewhere. Any suggestions where? Maybe a `utils` package? eg. `github.com/codeready-toolchain/member-operator/pkg/utils` with a `route.go` file?

File: pkg/che/tokencache.go (hunk 5 above)
Comment (rajivnathan) [reply pkg/che/tokencache.go:69]: Good suggestion, let me try splitting it up

File: pkg/che/tokencache.go (hunk 6 above)
Comment (rajivnathan) [reply pkg/che/tokencache.go:139]: 5 minutes

File: pkg/rest/rest.go (hunk 7 above)
Comment (rajivnathan) [reply pkg/rest/rest.go:21]: oh, I just copied it from https://github.com/codeready-toolchain/devcluster/blob/6defc6cce09762e3b65af412d20ba4970dd30456/pkg/rest/rest.go 🙂  and figured errors are ignored because it's mainly for cleanup

File: pkg/che/tokencache.go (hunk 5 above)
Comment (rajivnathan) [reply pkg/che/tokencache.go]: Done in 89b1e27e60e79ef25a5ba269afd473d55c6756eb

File: pkg/rest/rest.go (hunk 7 above)
Comment (alexeykazakov) [reply pkg/rest/rest.go:21]: yeah. we don't really care about errors there. We could log them though. Or at least make the linters happy by using _

File: pkg/che/tokencache.go [hunk 8]
```diff
@@ -0,0 +1,156 @@
...
+	if user == "" || pass == "" {
+		return TokenSet{}, fmt.Errorf("the che admin username and/or password are not configured")
+	}
+
+	// get keycloak URL
+	cheKeycloakURL, err := utils.GetRouteURL(cl, cfg.GetCheNamespace(), cfg.GetCheKeycloakRouteName())
+	if err != nil {
+		return TokenSet{}, err
+	}
+	log.Info("Che Keycloak Route", "URL", cheKeycloakURL)
```
[the remark this answers is not in the document]
Comment (rajivnathan) [reply pkg/che/tokencache.go:76]: ah, that's a really good point. I've addressed it in a0a013d320dc6dadea441b5e3afe7d52e9d972d6

File: pkg/rest/rest.go (hunk 7 above)
Comment (rajivnathan) [reply pkg/rest/rest.go:21]: Updated the `ReadBody` func to return the error and let the caller decide what to do.
Updated the `CloseResponse` func by using `_` to make the linters happy

**Review by alexeykazakov — APPROVED** [review]
Looks good!

File: pkg/che/tokencache.go (hunk 6 above)
Comment (MatousJobanek) [reply pkg/che/tokencache.go:146]: ok, then I would make the time cushion maybe a bit smaller - eg. 10seconds - or is there any reason why do we need to check that it will be valid at least one minute?

File: pkg/che/tokencache.go [hunk 9]
```diff
@@ -0,0 +1,163 @@
...
+
+// tokenExpired return false if the token is not nil and good for at least one more minute
+func tokenExpired(token *TokenSet) bool {
+	return token == nil || time.Now().After(time.Unix(token.Expiration-60, 0))
+}
+
+// newHTTPClient returns a new HTTP client with some timeout and TLS values configured
+func newHTTPClient() *http.Client {
+	var netTransport = &http.Transport{
+		Dial: (&net.Dialer{
```
Comment (MatousJobanek) [inline pkg/che/tokencache.go:152]: just a small detail - Dial is deprecated in http.Client - from the doc:
```
	// Deprecated: Use DialContext instead, which allows the transport
	// to cancel dials as soon as they are no longer needed.
	// If both are set, DialContext takes priority.
```

**Review by MatousJobanek — APPROVED** [review]
Looks good, thanks for addressing my comments :+1:

File: pkg/che/tokencache.go (hunk 9 above)
Comment (rajivnathan) [reply pkg/che/tokencache.go:152]: Thanks, I made a note of it. Will handle it in a following PR
